### ISSUE TITLE
Feature Request: avoid allocating logits buffer for encoder-only models

### ISSUE THREAD

**Issue by vschettino** [ggml-org/llama.cpp#29388]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

It is my understanding that `llama_context::output_reserve()` still fully allocates logits buffer for encoder-only models (BERT / XLM-RoBERTa: bge-m3, nomic-embed), which should never use those for their purpose. Local tests showing RAM memory growth compatible with `n_tokens` which leads me to believe it allocates the expected logits buffer by current logic (fix/design for #14204) 

```bash
=== bge-m3-f16.gguf
{"n_tokens":2000,"dim":1024}
n_vocab=250002  RSS before=796MiB after=2869MiB  growth=2074MiB
expected logits buffer n_vocab*N*4 = 1907MiB
=== nomic-embed-text-v1.5.f16.gguf
{"n_tokens":2000,"dim":768}
n_vocab=30522  RSS before=296MiB after=713MiB  growth=417MiB
expected logits buffer n_vocab*N*4 = 233MiB
```

My main hypothesis is that we really don't need that buffer in encoder-only models, so it would be nice to keep it free if that confirms.  

### Motivation

That memory is expensive (`(n_vocab * n_tokens * 4)` bytes) and from what I can see is RAM only, which makes it hard for smaller setups (homelab, consumer hardware).

### Possible Implementation

[#25964](https://github.com/ggml-org/llama.cpp/pull/25964) would be relevant here as an opt-in `--no-logits-for-embeddings` flag potentially would apply here. However I think there's a simpler fix by just checking if the model has logits in the first place:
- At init: llama_context[ already builds a reserve graph](https://github.com/ggml-org/llama.cpp/blob/master/src/llama-context.cpp#L676-L688) for prompt processing (`graph_reserve()`, then `get_gf_res_reserve()`. Right after that, record whether that graph has a logits output: `get_gf_res_reserve()->get_logits() != nullptr`.
- In [output_reserve():](https://github.com/ggml-org/llama.cpp/blob/master/src/llama-context.cpp#L676-L688) use that recorded flag instead of has_logits = true. Keep the T5 special case as it is.
